Experience Issues - Hide non template first themes behind a 'more themes' button.

Addison-Stavlo opened the pull request:
#### Changes proposed in this Pull Request

* Hide All Themes Search behind a 'More Themes' Button.
* Add a default list of recommended themes (template first) at the top of the Theme selection page.
* Adds a recommendedThemes section to the Themes store and corresponding action for retrieving recommended themes.

Note:  This PR **DOES NOT** introduce all of the style changes based on design mockups for the new themes section.  This address the main blocking issue of hiding old themes behind a button while introducing a list of recommended ones at the top.  Some of these design changes may require some significant restructuring of some of the components functionalities which would be best addressed in follow up PRs.

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Switch to this branch and run `npm start` in wp-calypso.
* Once built, navigate to calypso.localhost.
* Navigate to the Themes section and ensure the list of recommended Themes renders at the top of the page.
* Verify you can access and switch to these themes.
* Scroll down and click the "Show All Themes" button.
* Verify that the search functionality and selecting themes works as intended.
* Do this for both simple and atomic sites.

**Things to Consider**
Since there are so many special cases that can contribute to different components showing and not showing, its a bit hard to tell if everything is in the correct place.  If we can think of more special cases that would cause the Themes section to display differently, please note them and we can test those as well.
* Are we hiding any components behind this button that we do not want hidden? (Query Theme Filters, Upwork Banner, Upsell Banner, etc.)
* Are we showing any components that should be hidden?  (Page View Tracker, Sub Masterbar Nav, etc.)
* Are there any other things we need to test or consider for this?  This is a section of the codebase that has a lot of "moving pieces" and specialized functionality while the impact of this change would be live to all calypso users.

**Before**
![Screen Shot 2019-11-21 at 7 34 13 PM](https://user-images.githubusercontent.com/28742426/69388126-f1866a80-0c95-11ea-82e3-dcd3694eeb1b.png)


**After - Top Of Page**
![Screen Shot 2019-11-21 at 7 16 12 PM](https://user-images.githubusercontent.com/28742426/69387327-7b810400-0c93-11ea-8f1a-d1599ffbf659.png)

**After - After Clicking Button**
![Screen Shot 2019-11-21 at 7 16 35 PM](https://user-images.githubusercontent.com/28742426/69387349-92bff180-0c93-11ea-96a3-fd09b74a709e.png)

Note that the above screenshot is on an Atomic site and that the 'Uploaded Themes' section is below our Recommended themes.  It was mentioned in #37285 that 'Uploaded Themes' may be best shown above 'Recommended Themes'.  However, the component that renders this "Uploaded Themes" section in the atomic site happens to be the exact same component call used to show all wpcom search results for simple sites (then in the case of atomic the wpcom search results come from an extra child being passed in as props).  

Thus to get this 'Uploaded Themes' section above "Recommended Themes" for atomic sites while hiding the normal wpcom search results for simple sites is one of the changes that may take a bit of restructuring and would be best addressed in a follow up PR if necessary.  I would however argue that below the search bar is a fitting place for these 'Uploaded Themes' since this section is designed to filter based off of the search bar input.
  
Fixes #37285

Addison-Stavlo committed:
Hid showcase behind button

Addison-Stavlo committed:
showing varia themes above button

Addison-Stavlo committed:
Add/hide non template first themes try themeselection (#37648)

* using themeSelection component instead

* removed unused var

* onclick and modal working

Addison-Stavlo committed:
removed getThemes out of CDM

Addison-Stavlo committed:
pagination removed from ThemeSelection used in RecommendedThemes

Addison-Stavlo committed:
reorganized existing components and classNames to better fit original layout

Addison-Stavlo committed:
query updated

Addison-Stavlo committed:
button centered and translated

Addison-Stavlo committed:
loading spinner added, pagination function conditioned

Addison-Stavlo committed:
ref added for scroll to location after page() is called

Addison-Stavlo committed:
moved theme retrieval to store

Addison-Stavlo committed:
Merge branch 'master' of https://github.com/Automattic/wp-calypso into add/hide-non-template-first-themes

Addison-Stavlo committed:
moved chilren into hiding for atomic sites

Addison-Stavlo committed:
extra content now display hidden instead of conditionally rendered, fixes inconsistencies between atomic

Addison-Stavlo committed:
removed an empty line, ooo

Addison-Stavlo committed:
updated comment style

noahtallen commented on client/my-sites/themes/theme-showcase.jsx:
```diff
@@ -284,8 +285,69 @@ class ThemeShowcase extends React.Component {
-					{ this.props.children }
+
+					<div className="theme-showcase__open-showcase-button-holder">
+						{ isShowcaseOpen ? (
```
Should this be separated into a new component, and then we show `RecommendedThemes` along with `AllThemes`?

that could also be used if we need to gate the feature to certain users

Addison-Stavlo replied:
Perhaps it should.  These were all in the ThemeShowcase from the beginning, and not being 100% certain which components may require being hidden or extracted as part of `AllThemes` I haven't taken any steps in that direction yet.  

It's currently also the case that `RecommendedThemes` requires the `AllThemes` part of the code to run to maintain all of its functionality.  This is why I am hiding the `AllThemes` section with `display: none` instead of conditionally loading it in the first place.  When I had it conditioned to not load until the button is pushed, `RecommendedThemes` would have small bugs in its functionality until `AllThemes` was opened and loaded.

I am going to investigate a bit more to try and clean this all up, and in that case creating the `AllThemes` component could be good as well.  At its current state it works for how we want it to work (alongside `AllThemes`), but is lacking in the sense that `Recommended Themes` is not completely stand-alone and would cause issues down the line if we tried to use it somewhere without the `AllThemes` portion.

noahtallen commented on client/my-sites/themes/themes-selection.jsx:
```diff
@@ -144,11 +150,15 @@ class ThemesSelection extends Component {
 				<ThemesList
 					upsellUrl={ upsellUrl }
-					themes={ this.props.themes }
+					themes={ this.props.recommendedThemes || this.props.themes }
```
Why do we need to pass both  `recommendedThemes` and `themes` into the component? Is there a way to abstract it so that the component only takes a `themes` prop, and then `recommendedThemes` are given to that prop in the right context? (I may also be confused about what this does :)

Addison-Stavlo replied:
Here we are passing one of the two into the `ThemesList` component as the `themes` prop.  So if we passed `recommendedThemes` into `ThemesSelection` like we do from the `RecommendedThemes` component, then `ThemesList` receives that as the `themes` prop.  If there is no `recommendedThemes` prop (how the component is normally used), then it falls back to pass `props.themes` connection from state.  So in the case of `ThemesList` it is always receiving only one of those as the `themes` prop.   

Or were you saying that the the `ThemesSelection` component itself should be cleaned up in a way that we aren't passing the recommended list down?  I suppose we could make this a condition in the `connect` HOC to determine what `themes` should be so we don't need to have `recommendedThemes` inside the actual component class.  But we do find `recommendedThemes` helpful as a boolean check in the component class to condition things like `incrementPage` function and whether or not we should render certain other components.  So Im not sure if thats going to be very helpful either. 🤔

noahtallen commented on client/state/themes/actions.js:
```diff
@@ -913,3 +916,27 @@ function suffixThemeIdForInstall( state, siteId, themeId ) {
+			search: '',
+			number: 50,
+			tier: '',
+			filter: 'auto-loading-homepage',
```
Noice!

noahtallen commented on client/state/themes/reducer.js:
```diff
@@ -459,6 +462,28 @@ export const themeFilters = withSchemaValidation( themeFiltersSchema, ( state =
+ * @param  {object} action Action payload
+ * @returns {object}        Updated state
+ */
+const recommendedThemes = ( state = { isLoading: false, themes: [] }, action ) => {
```
I think we usually try to add tests for new parts of the state. You can probably find examples by looking at other reducers :)

Addison-Stavlo replied:
Great suggestion.  I have added some basic tests for the actions and reducers.

noahtallen reviewed (commented):
😍 This looks great! It works very smoothly for me, with no functionality problems. I do have a couple thoughts, and I hope we get a lot of eyes on this change, since it is a big one and affects a lot of users :)

I imagine the tests are failing because some of them probably test that this themes page looks a certain way, so those should probably be updated to match the new format.


- I notice that when going to "theme info" and then back to the theme list, the recommended themes re-loads. I wonder if we have to run `getRecommendedThemes` every time, since this list does not change very often. Or, we could use the existing themes if they already exist in the state instead of showing the loading indicator -- and then when the query finishes, it can update the state behind the scenes.
- Should we gate this change to a subset of users? It seems like one of the largest changes in our experience issues, so I wonder if we need to contain this change to _only_ people who have FSE-eligible sites. (Note that [this selector can be used](https://github.com/Automattic/wp-calypso/blob/f2c9d99b07e14ea938829e9b2e5b31b6fb388a90/client/state/selectors/is-site-using-full-site-editing.js) to detect that with some tweaks. This would work out very will once all of the themes shown here have FSE support.)
- Should we duplicate "recommended themes" in the "all themes" list? It feels weird that the first screen-ful of themes is the same as the ones we were already looking at.

Most of the above, I don't know the answer to, so it'll be good to get more feedback :)

apeatling commented:
Thanks for your work on this one @Addison-Stavlo! Couple of things:

<img width="1015" alt="Screen Shot 2019-11-24 at 8 43 49 AM" src="https://user-images.githubusercontent.com/1464705/69497937-05e58580-0e97-11ea-87f6-57643215f7f1.png">

cc @shaunandrews for the site title + blurb bit.

shaunandrews commented:
Maybe something like this:

<img width="750" alt="image" src="https://user-images.githubusercontent.com/191598/69556071-233b5200-0f72-11ea-99dd-2defb6adc636.png">

Addison-Stavlo committed:
restrict api call for recommendedThemes to only happen if they are not already on state

Addison-Stavlo committed:
added heading description to Advanced Themes section

Addison-Stavlo reviewed (commented):
@shaunandrews  that looks great.  Is there similar descriptive sentence you would like to add just below "Recommended Themes" in the section above as well?  "These are our latest themes that support the most customization for new users" or something?

Addison-Stavlo committed:
filtered recommendedThemes out of default query

Addison-Stavlo commented:
@apeatling - 
I just updated it to filter the recommended themes out of the search input by default.  I noticed that filtering them out in all cases may cause some un-wanted issues as seen below when people may try to search via the "auto-loading-homepage" tag or search a theme by its specific name:

![Screen Shot 2019-11-25 at 3 17 25 PM](https://user-images.githubusercontent.com/28742426/69574822-0e24ea00-0f97-11ea-8336-c4c770ba45ec.png)

 or
 
![Screen Shot 2019-11-25 at 3 17 41 PM](https://user-images.githubusercontent.com/28742426/69574808-07967280-0f97-11ea-9621-abd9b7d3c989.png)

To avoid this I have made it so that this will not show recommended themes in the "all themes" section when it is initially opened or reverts back to a blank input/query, but they will show up once the input starts looking for specific results.  I am assuming this/below is the behavior we are looking for?

![Screen Shot 2019-11-25 at 3 31 49 PM](https://user-images.githubusercontent.com/28742426/69575775-e898e000-0f98-11ea-99b7-aad73ec17341.png)

![Screen Shot 2019-11-25 at 3 32 11 PM](https://user-images.githubusercontent.com/28742426/69575764-e59def80-0f98-11ea-9295-628848bf8921.png)

![Screen Shot 2019-11-25 at 3 31 18 PM](https://user-images.githubusercontent.com/28742426/69575779-eafb3a00-0f98-11ea-8369-3743c5cac550.png)

Addison-Stavlo reviewed (commented):
@noahtallen 

>I notice that when going to "theme info" and then back to the theme list, the recommended themes re-loads. 

Good catch!  I have updated this and it should no longer trigger any fetch if this is already on state.

> I imagine the tests are failing because some of them probably test that this themes page looks a certain way, so those should probably be updated to match the new format.

Definitely!  I was holding off at first for some feedback so I knew we were on the right direction but I think its safe to update these now.  Will do shortly.

apeatling commented:
Tested this out, and I think this is mergeable from an experience point of view. One nitpick, remove the colon at the end of the Recommended Themes title.

noahtallen commented:
> I think this is mergeable from an experience point of view.

I think it'd be good to write a happiness update about the changes since it is big :)

Addison-Stavlo committed:
removed colon

Addison-Stavlo commented:
@noahtallen 
> I think it'd be good to write a happiness update about the changes since it is big :)

Definitely a big yes!  I am planning to write a p2 about it before merging for visibility purposes (is a happiness update different or just an update on the happiness p2?).  

@apeatling 
> Tested this out, and I think this is mergeable from an experience point of view. 

Thats great!  I'l update the tests and do a little bit more digging to see if making these components a bit more stand-alone friendly will be a relatively simple fix, but I am glad to know from a experience PoV that this is about ready 😄 

> One nitpick, remove the colon at the end of the Recommended Themes title.

Done.

Addison-Stavlo committed:
updated filtering to be more specific to recommendedThemes state

apeatling commented:
@Addison-Stavlo We need to add a tracks event on the load more button so we can track this in funnels. /cc @ianstewart

Addison-Stavlo commented:
It looks like this also effects the logged out themes view such as `WordPress.com/themes` or `calypso.localhost:3000/themes`.  So my first focus today is going to be investigating and making changes as needed.  

In terms of style and functionality for this "Logged Out" themes area, it is working the same as the logged in user section but with its default filter bar above everything (see below).  I am thinking we may want to remove this Recommended Themes section from the logged out view?  @shaunandrews @apeatling how do we feel about this?

![Screen Shot 2019-11-27 at 11 39 24 AM](https://user-images.githubusercontent.com/28742426/69742342-a0013400-110a-11ea-8820-7ed6f942b0de.png)

As seen in the screenshot above, the top blue bar applies filters to the themes shown in the "Advanced Themes" section that is hidden behind the button and much further down the page once opened.  So the section of themes that the blue bar filters is so far down the page that you cant see them when you are using the blue bar.

Also, does anybody know of any other areas where a logged out user can view the Theme Showcase?

Addison-Stavlo commented:
> @Addison-Stavlo We need to add a tracks event on the load more button so we can track this in funnels. /cc @ianstewart

Ok.  Il be happy to add it but I am not totally sure how/what I am adding.  I will look into this more, but do we have any more info or insight on this?

Addison-Stavlo committed:
set default loading for rec themes state to true

Addison-Stavlo committed:
updated selectors tests for getThemesForQueryIgnoringPage

Addison-Stavlo committed:
updated action and wrote action tests

Addison-Stavlo committed:
added tests for reducer

apeatling commented:
For tracks event, I think you need to call `recordTracksEvent` from https://github.com/Automattic/wp-calypso/tree/master/client/state/analytics when the button is clicked.

The event can be called `calypso_themeshowcase_more_themes_clicked`

Addison-Stavlo committed:
updated conditional rendering for the loggOut version of the component

Addison-Stavlo committed:
added track event for more themes click

Addison-Stavlo commented:
I was able to update the logged out area not to render the `recommendedThemes` section and to have the old themes search open by default.  So this area now seems to look and function as the live version of `wordpress.com/themes` currently does.

I have also been able to add tests for actions/reducers, get the LoggedOut component tests to pass again, and have added the track event to the "More Themes" button.

apeatling commented:
This is all looking good to me. :+1:

vindl reviewed (approved):
This worked as described for me. 👍

vindl commented:
I've looked into some of the test failures below, and it seems like they are related to the changes here: `[WPCOM] Activating Themes...`.

noahtallen reviewed (approved):
It's working well for me! Should be ready to go once the tests are passing :)

Addison-Stavlo commented:
Hrm, you are right.. while all the standard tests from `npm run test` seem to pass, it looks like there are some e2e tests here failing and they all say `Timed out waiting for element with css selector of 'a[data-e2e-value="free"]' to be clickable`.  I wonder if this has something to do with us hiding the search bar by default...

Addison-Stavlo committed:
testing e2e test update?

Addison-Stavlo committed:
more update on e2e

Addison-Stavlo commented:
After adding a step in a couple parts of the e2e test to open the button, it looks like there is 1 theme related test not passing now.  It looks like it is coming from here:

https://github.com/Automattic/wp-calypso/blob/d8bf6a036301628802e0853eb771ddfb74e62b3b/test/e2e/specs/wp-theme-multisite-spec.js#L70-L72

Addison-Stavlo commented:
I did a couple small updates to the e2e tests so that the "more themes" button is clicked where the tests need access to the hidden content.  There is 1 more themes related test not passing and it looks like it is coming from here:
https://github.com/Automattic/wp-calypso/blob/d8bf6a036301628802e0853eb771ddfb74e62b3b/test/e2e/specs/wp-theme-multisite-spec.js#L70-L72

It is timing out waiting for this "Try & Customize" button to appear and be clickable.  I am thinking this accesses the first theme option on the page and tries to do this, but since the first theme option is now a template-first theme (which don't appear to have that option) it times out waiting for a button that will never appear.

I will look into this more tomorrow.  But if that is the case, what are our thoughts about this test?  Is `Try & Customize` something that we are phasing out as we move to FSE?  Is this a section of the themes e2e tests that should be removed?  Or do we work at updating this in some manner that it accesses a non template-first-theme in the bottom section of the screen instead?

noahtallen commented:
> it times out waiting for a button that will never appear.

Really interesting because the tests did seem to work when we changed that button, and the first theme was a template first theme in that case too. I wonder if clicking "show more" first will make it work? Or if there's a way to target a non-template first theme?

For legacy purposes, it could be good to make the test continue working for legacy themes, but also add a similar test for the template first themes

Addison-Stavlo committed:
e2e added openShowcase to clickPopoverItem

Addison-Stavlo committed:
added data tag to hr button replacement to try to better support e2e

Addison-Stavlo committed:
removed data ref from hr and moved openshowCase calls

Addison-Stavlo committed:
added not selector to e2e test and classname to recommended Themes

Addison-Stavlo committed:
added classname to all themes category for e2e selection

Addison-Stavlo commented:
At this point I have fixed most of the e2e tests.  Currently there is 1 failing, saying "Successful activation dialog should show the correct theme in the current theme bar".  The test seems to activate a different theme and then check the Current Theme Bar displays the proper theme.  The test fails and sees the old theme's name present.  However, when I try to do replicate this I am not seeing any issues.  🤔

vindl commented:
> At this point I have fixed most of the e2e tests. Currently there is 1 failing, saying "Successful activation dialog should show the correct theme in the current theme bar". The test seems to activate a different theme and then check the Current Theme Bar displays the proper theme. The test fails and sees the old theme's name present. However, when I try to do replicate this I am not seeing any issues. 🤔

Looping in @Stojdza for help with e2e tests. :)

Stojdza committed:
More specific selector - get first theme from .theme-showcase__all-themes

Stojdza commented:
Failing tests activated correct theme (same as before), but it was expecting another theme - a wrong one. This happened because [test was getting the first theme](https://github.com/Automattic/wp-calypso/pull/37696/files?file-filters%5B%5D=.js#diff-6af1f5cc87189fb61b8e9d8310cb7255L65) from recommended Themes list, introduced in this PR. 

I've [updated the selector](https://github.com/Automattic/wp-calypso/pull/37696/commits/7e5af6eeadd966f89606daed262fdc67d0c310da) to be more specific. Now, we are getting the first theme from all themes list, at the bottom of the page.

Copons commented:
Noting that `ci/circleci: lint-and-translate` seems to fail at lines unrelated to this change.

Thanks @Stojdza for helping us here! I was about to giving it a deeper look, so you spared me what could have been a long investigation 😄 

I've restarted the e2e, finger crossed!

Addison-Stavlo commented:
@Stojdza - Thank you so much for the help!

> This happened because test was getting the first theme from recommended Themes list, introduced in this PR.

Ok that makes sense now.  I had some other tests failing for a similar reason earlier, I missed that this was the same thing happening here.

> I've updated the selector to be more specific. Now, we are getting the first theme from all themes list, at the bottom of the page.

Thanks again!

apeatling commented:
🎉Nice work @Addison-Stavlo !

vindl commented:
Thanks for helping us with e2e test @Stojdza and great work once again @Addison-Stavlo! 🚀
